File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.py
class Solution:
    def myAtoi(self, s: str) -> int:
        INT_MAX = 2**31 - 1
        INT_MIN = -2**31
        index=0
        ans=0
        n=len(s)
        while(index<n and s[index]==" "):
            index+=1
        if(index==n):
            return 0
        sign=1
        if(s[index]=="-"):
            sign=-1
            index+=1
        elif(s[index]=="+"):
            sign=1
            index+=1

        
        print(index)
        while(index<n and s[index].isdigit()):
            digit=int(s[index])

            if( ans > (INT_MAX - digit ) // 10 ):
                return INT_MAX if sign==1 else INT_MIN
            ans=(ans*10)+int(digit)
            index+=1



        return ans*sign
        
        

File: 0008-string-to-integer-atoi/0008-string-to-integer-atoi.java
class Solution {
    // static final int INT_MAX=2147483648;
    // static final int INT_MIN=-2147483648;
    private static int helper(String s, int i, long ans, int sign) {
        if (i >= s.length() || !Character.isDigit(s.charAt(i))) {
            return (int) (sign * ans);

        }
        int val = s.charAt(i) - '0';
        if (sign == 1) {
            if (ans > Integer.MAX_VALUE / 10 || (ans == Integer.MAX_VALUE / 10 && val > 7))
                return Integer.MAX_VALUE;
        } else {
            if (ans > Integer.MAX_VALUE / 10 || (ans == Integer.MAX_VALUE / 10 && val > 8))
                return Integer.MIN_VALUE;
        }
        ans = (ans * 10) + val;
        return helper(s, i + 1, ans, sign);
    }

    public int myAtoi(String s) {
        if(s.length()==0) return 0;
        int i = 0;
        while (i < s.length() && s.charAt(i) == ' ')
            i++;
        int sign = 1;
        if(i>=s.length()) return 0;
        if (s.charAt(i) == '-' || s.charAt(i) == '+') {
            if (s.charAt(i) == '-')
                sign = sign * -1;
            i++;
        }

        return helper(s, i, 0, sign);

    }
}